### PR TITLE
Fail Kubelet at startup if swap is configured with cgroup v1

### PR DESCRIPTION
**Pull request by iholder101**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Follow-up to https://github.com/kubernetes/kubernetes/pull/118764 and https://github.com/kubernetes/kubernetes/issues/119726.

Since swap is no longer supported for cgroup v1, fail kubelet if it starts with swap and cgroup v1. This way, in future versions of k8s, we could turn NodeSwap on by default.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Fail Kubelet at startup if swap is configured with cgroup v1
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md
```

### CODE REVIEW

File: pkg/kubelet/cm/container_manager_linux.go [hunk 1]
```diff
@@ -225,6 +225,10 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
 				return nil, fmt.Errorf("running with swap on is not supported, please disable swap! or set --fail-swap-on flag to false. /proc/swaps contained: %v", swapLines)
 			}
 		}
+	} else {
+		if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) && !cgroups.IsCgroup2UnifiedMode() {
+			return nil, fmt.Errorf("running swap with cgroups v1 is not supported. please either disable %s feature gate", kubefeatures.NodeSwap)
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/cm/container_manager_linux.go:230]:
```suggestion
			return nil, fmt.Errorf("running swap with cgroups v1 is not supported. Please either disable %s feature gate or start using cgroup v2.", kubefeatures.NodeSwap)
```

File: pkg/kubelet/cm/container_manager_linux.go [hunk 2]
```diff
@@ -225,6 +225,10 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
 				return nil, fmt.Errorf("running with swap on is not supported, please disable swap! or set --fail-swap-on flag to false. /proc/swaps contained: %v", swapLines)
 			}
 		}
+	} else {
+		if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) && !cgroups.IsCgroup2UnifiedMode() {
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/cm/container_manager_linux.go:229]: I wasn't thinking about it before, but now I realize that the problem with this is that all tests running today with `--fail-on-swap=false` will stop working as the feature gate will be enabled by default.

Even though it is a desirable behavior to fail fast long term, breaking many scenarios is not ideal.

One option is to change the default value for the feature gate.

Comment (SergeyKanzhelev) [reply pkg/kubelet/cm/container_manager_linux.go:229]: + @deads2k (PRR reviewer for the SWAP KEP) for the opinion

Comment (iholder101) [reply pkg/kubelet/cm/container_manager_linux.go:229]: > I wasn't thinking about it before, but now I realize that the problem with this is that all tests running today with --fail-on-swap=false will stop working as the feature gate will be enabled by default.

Yes this is exactly right, and was one of the reasons I asked to do so in a follow-up :)

I think you're right, and that the right approach moving forward is:
1. Introduce a PR that would adapt the tests.
2. Merge this PR after making sure it won't break any lanes.

> One option is to change the default value for the feature gate.

What do you mean by that? Isn't the default behavior for the feature gate is to be disabled?
IIUC all we need to do is remove all jobs that are defined with cgroup v1 + NodeSwap enabled. We should keep (and maybe add more? but outside of this PR's scope) the jobs with swap + v2.

WDYT?

File: pkg/kubelet/cm/container_manager_linux.go (hunk 1 above)
Comment (iholder101) [reply pkg/kubelet/cm/container_manager_linux.go:230]: Thanks :) Probably my mind had a context switch and didn't switch back..

**Comment by iholder101** [comment]
/test all

Let's see exactly which tests fail to discuss our next steps.

File: pkg/kubelet/cm/container_manager_linux.go (hunk 1 above)
Comment (pacoxu) [reply pkg/kubelet/cm/container_manager_linux.go]: This is not addressed yet but outdated for rebasing.

**Comment by iholder101** [comment]
@SergeyKanzhelev ping
would you mind having another look?

File: pkg/kubelet/cm/container_manager_linux.go [hunk 3]
```diff
@@ -224,6 +224,8 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
 				return nil, fmt.Errorf("running with swap on is not supported, please disable swap! or set --fail-swap-on flag to false. /proc/swaps contained: %v", swapLines)
 			}
 		}
+	} else if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) && !cgroups.IsCgroup2UnifiedMode() {
+		return nil, fmt.Errorf("running swap with cgroups v1 is not supported. please either disable %s feature gate", kubefeatures.NodeSwap)
```
Comment (kannon92) [inline pkg/kubelet/cm/container_manager_linux.go:228]:
```suggestion
		return nil, fmt.Errorf("running swap with cgroups v1 is not supported. please disable %s feature gate", kubefeatures.NodeSwap)
```

**Comment by kannon92** [comment]
@iholder101 thank you for this change. I moved the code to https://github.com/kubernetes/kubernetes/pull/122241 so we can continue this work.

/close

**Pull request closed without merging**